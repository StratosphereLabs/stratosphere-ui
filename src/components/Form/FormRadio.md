# FormRadio

The `FormRadio` component is a wrapper around the `Radio` component from DaisyUI. It is used to render a group of radio buttons that are part of a form, and integrates with `react-hook-form` for managing form state.

## Props

The `FormRadio` component accepts the following props:

- `className?: string`:
  Optional class name to be applied to the outermost container div element.
- `color?: string`:
  The color of the radio buttons. If not provided, the default is 'ghost'.
- `controllerProps?: Omit<UseControllerProps<Values>, 'name'>`:
  Optional props for the `useController` hook from `react-hook-form`, which is used to integrate the radio buttons with the form state.
- `inputRef?: RefObject<HTMLInputElement>`:
  Optional ref object to be passed to the underlying input component.
- `isRequired?: boolean`:
  Whether the field is required. If true, an asterisk will be added to the label text.
- `labelText?: string`:
  Optional text to be used for the label of the radio button group.
- `name: Path<Values>`:
  The name of the field, used as the key for the form state object in `react-hook-form`.
- `options: RadioOption[]`:
  An array of objects containing the id, label, and value of each radio button option.
- `showDirty?: boolean`:
  If true, the field will be styled differently if it has been modified by the user.

## Usage

To use the `FormRadio` component, simply import it into your React component and render it with the appropriate props:

```tsx
import { useForm } from 'react-hook-form';
import { Form, FormRadio, RadioOption } from 'stratosphere-ui';

interface FormValues {
  gender: string;
}

const genderOptions: RadioOption[] = [
  { id: 'male', label: 'Male', value: 'male' },
  { id: 'female', label: 'Female', value: 'female' },
];

export const MyForm = () => {
  const methods = useForm<FormValues>();
  const onSubmit = (data: FormValues) => {
    console.log(data);
  };
  return (
    <Form methods={methods} onSubmit={onSubmit}>
      <FormRadio
        name="gender"
        options={genderOptions}
        controllerProps={{ control }}
        labelText="Gender"
      />
    </Form>
  );
};
```
