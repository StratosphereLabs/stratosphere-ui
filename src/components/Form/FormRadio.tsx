import classNames from 'classnames';
import { HTMLProps } from 'react';
import { FieldValues, useController } from 'react-hook-form';

import { useFieldColor } from '../../hooks';
import { FormError } from './FormError';
import { FormLabelText } from './FormLabelText';
import { FormFieldProps } from './types';

export interface RadioOption {
  className?: string;
  id: string | number;
  label: string;
  value: string;
}

export const RADIO_COLORS = [
  'neutral',
  'primary',
  'secondary',
  'accent',
  'success',
  'warning',
  'info',
  'error',
] as const;

export const RADIO_SIZES = ['xl', 'lg', 'md', 'sm', 'xs'] as const;

export type RadioColor = (typeof RADIO_COLORS)[number];

export type RadioSize = (typeof RADIO_SIZES)[number];

export interface FormRadioProps<Values extends FieldValues>
  extends Omit<FormFieldProps<Values>, 'placeholder'>,
    Omit<HTMLProps<HTMLInputElement>, 'name' | 'size'> {
  color?: RadioColor;
  inputClassName?: string;
  options: RadioOption[];
  size?: RadioSize;
}

export const FormRadio = <Values extends FieldValues>({
  className,
  color,
  controllerProps,
  inputClassName,
  isRequired,
  labelText,
  name,
  options,
  showDirty,
  size,
  ...props
}: FormRadioProps<Values>) => {
  const {
    field: { value, ...field },
    fieldState: { error },
  } = useController({
    name,
    ...controllerProps,
  });
  const fieldColor = useFieldColor(name, showDirty);
  const currentColor = fieldColor ?? color ?? undefined;
  return (
    <div className={classNames('form-control', className)}>
      {labelText !== undefined ? (
        <div className="label">
          <FormLabelText isRequired={isRequired}>{labelText}</FormLabelText>
        </div>
      ) : null}
      {options.map(
        ({ className: optionClassName, id, label, value: optionValue }) => (
          <label
            key={id}
            className={classNames('label cursor-pointer', optionClassName)}
          >
            <span className="label-text">{label}</span>
            <input
              {...field}
              {...props}
              checked={value === optionValue}
              className={classNames(
                'radio',
                currentColor && `radio-${currentColor}`,
                size && `radio-${size}`,
                inputClassName,
              )}
              type="radio"
              value={optionValue}
            />
          </label>
        ),
      )}
      {error?.message !== undefined ? (
        <FormError>{error.message}</FormError>
      ) : null}
    </div>
  );
};
