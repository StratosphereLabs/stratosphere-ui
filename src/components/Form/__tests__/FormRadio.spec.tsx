import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FormProvider from './FormProvider';
import { FormRadio, RadioOption } from '../FormRadio';
import { zodResolver } from '@hookform/resolvers/zod';
import { z } from 'zod';

const options: RadioOption[] = [
  { id: 1, label: 'Option 1', value: 'option1' },
  { id: 2, label: 'Option 2', value: 'option2' },
];

describe('FormRadio', () => {
  it('renders radio buttons with the provided options', () => {
    render(
      <FormProvider defaultValues={{ radio: null }}>
        <FormRadio labelText="Test Radio" name="radio" options={options} />
      </FormProvider>,
    );
    options.forEach(option => {
      const radio = screen.getByLabelText(option.label);
      expect(radio).toBeInTheDocument();
      expect(radio).toHaveAttribute('value', option.value);
    });
  });

  it('sets the value of the selected radio button', async () => {
    const { rerender } = render(
      <FormProvider defaultValues={{ radio: null }}>
        <FormRadio name="radio" options={options} />
      </FormProvider>,
    );
    const option1 = screen.getByLabelText('Option 1');
    await userEvent.click(option1);
    expect(option1).toBeChecked();
    expect(option1).toHaveAttribute('value', 'option1');
    rerender(
      <FormProvider defaultValues={{ radio: null }}>
        <FormRadio name="radio" options={options} />
      </FormProvider>,
    );
    expect(screen.getByLabelText('Option 1')).toBeChecked();
  });

  it('renders an error message when an error is present', async () => {
    render(
      <FormProvider
        mode="onChange"
        defaultValues={{ radio: null }}
        resolver={zodResolver(
          z.object({ radio: z.string().min(10, 'Too short') }),
        )}
      >
        <FormRadio name="radio" options={options} />
      </FormProvider>,
    );
    await userEvent.click(screen.getByLabelText('Option 1'));
    await waitFor(() =>
      expect(screen.getByText('Too short')).toBeInTheDocument(),
    );
  });
});
