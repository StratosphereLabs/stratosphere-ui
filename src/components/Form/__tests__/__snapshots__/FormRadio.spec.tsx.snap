// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormRadio > renders radio buttons with the provided options 1`] = `
<div>
  <fieldset
    class="fieldset py-0"
  >
    <legend
      class="fieldset-legend"
    >
      Test Radio
       
    </legend>
    <div
      class="flex cursor-pointer items-center justify-between"
    >
      <span>
        Option 1
      </span>
      <input
        class="radio"
        name="radio"
        type="radio"
        value="option1"
      />
    </div>
    <div
      class="flex cursor-pointer items-center justify-between"
    >
      <span>
        Option 2
      </span>
      <input
        class="radio"
        name="radio"
        type="radio"
        value="option2"
      />
    </div>
  </fieldset>
</div>
`;
